PIPELINE: Abdomen pipeline

UdomkarnBoonyaprasert opened the pull request:
Adding pipeline to generate glb 3D models. Please note that in order to run the pipeline the segmentation container must be running. The functionality of this pipeline includes: 
- converting Dicom to Nifti
- resize and crop the Nifti (this is actually done during the Numpy array stage using `compNumpyTransformation`) to fit the max resolution for the segmentation model
- generate glb for each organs in the segmentation (8 in total)

NOTE: files copied from `PIPELINE/abdomen_with_resize_function_pipeline` branch to clean up commits

UdomkarnBoonyaprasert committed:
PIPELINE: update dicom2numpy
- update dicom2numpy with additional function to reorientate the image in numpy array form
- add a new dicom2nifti component which will be replacing the existing one

UdomkarnBoonyaprasert committed:
PIPELINE: add abdomen pipeline
- add new pipline: abdomenDicom2glb
- update pipelineList
- update function in compJobStatus to handle POST request with file

UdomkarnBoonyaprasert committed:
PIPELINE: fix issues from codeFactor

UdomkarnBoonyaprasert committed:
PIPELINE: fix commits

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -79,10 +79,19 @@ def resample(dataPath, new_spacing=[1, 1, 1]):
 
 
+def reorientateNumpyList(numpyList):
+    # tanspose numpy i.e. (z, y, x) ---> (x, y, z)
```
typo transpose

UdomkarnBoonyaprasert replied:
typo fixed

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,19 @@
+def sendFilePostRequest(url, inputFile, outputFile):
+    inputFile = str(pathlib.Path(inputFile))
+    outputFile = str(pathlib.Path(outputFile))
+    file = {"file": open(inputFile, "rb")}
```
use the `with` pattern for files. It helps with auto closing the file once it goes out of scope

UdomkarnBoonyaprasert replied:
changed to `with open(inputFile, "rb") as inputFileData:`

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,19 @@
+    inputFile = str(pathlib.Path(inputFile))
+    outputFile = str(pathlib.Path(outputFile))
+    file = {"file": open(inputFile, "rb")}
+    response = requests.post(url, files=file)
```
what happens if an error happens here.

boonwj replied:
e.g. timeout etc

UdomkarnBoonyaprasert replied:
Added timeout for post request and exception( sys.exit() as there's no point in continuing the pipeline)

boonwj commented on HoloPipelines/pipelines/pipelineAbdomenDicom2glb.py:
```diff
@@ -0,0 +1,26 @@
+def main(inputDicomPath, outputGlbFolderPath):
+    generatedNiftiPath = compDicom2nifti.main(inputDicomPath, "_temp.nii")
+    segmentedNiftiPath = compHttpRequest.sendFilePostRequest(
+        "http://localhost:5000/model",
```
change this to read from a env variable

UdomkarnBoonyaprasert replied:
this pipeline now takes url as an argument

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,19 @@
+    file.write(response.content)
+    file.close()
+    returnCode = response.status_code
+    print(returnCode)
```
What do you use the return code for?

boonwj replied:
use it for error handling?

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,47 @@
+    highest = max(x, y, z)
+    if highest > limit:
+        rezieRatio = limit / highest
+        img = resize(img, rezieRatio, rezieRatio, rezieRatio)
```
resizeRatio typo.

UdomkarnBoonyaprasert replied:
typo fixed

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,47 @@
+import scipy.ndimage
+
+
+def crop(img, cropx, cropy, cropz):
```
name is misleading as this does centering and not just cropping the edges.

boonwj replied:
change crop[x,y,z] to something more explicit res[x,y,z]

UdomkarnBoonyaprasert replied:
renamed to res()

boonwj replied:
ah i see where the confusion came about. My bad i was initially referring to the arguments.

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def resize(img, ratioX, ratioY, ratioZ):
+    return scipy.ndimage.interpolation.zoom(img, [ratioX, ratioY, ratioZ])
```
If this one line function is used only once in this file. It can be called solely on its own.

UdomkarnBoonyaprasert replied:
removed resize()

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,47 @@
+def sizeLimit(img, limit):
+    if len(img.shape) == 5:
+        x, y, z, t1, t2 = img.shape
+    else:
```
making an assumption that "img" will be of size 5 or 3. Perhaps handle other incorrect usages of this function.

UdomkarnBoonyaprasert replied:
update to:
- check if len >= 3
- read the first 3 value from `img.shape`

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,47 @@
+
+def sizeLimit(img, limit):
+    if len(img.shape) == 5:
+        x, y, z, t1, t2 = img.shape
```
Also, t1 and t2 are never used. So perhaps all you need are the first 3 values

UdomkarnBoonyaprasert replied:
t1 and t2 removed

boonwj commented on HoloPipelines/pipelines/components/compSeparateNumpy.py:
```diff
@@ -0,0 +1,54 @@
+    else:
+        try:
+            # if inputNumpy is path to npy file
+            originalNumpy = np.load(str(pathlib.Path(inputNumpy)))
```
it's kinda wierd that your input can handle both the numpy array and also a numpy file. Just stick to one and make sure others use it correctly

UdomkarnBoonyaprasert replied:
component now only accepts NumPy array

boonwj commented on HoloPipelines/pipelines/components/compSeparateNumpy.py:
```diff
@@ -0,0 +1,54 @@
+            singleHUnumpy = singleHUnumpy.astype(int)
+            makeObj.main(
+                singleHUnumpy,
+                0,
```
use the flag= for this argument

UdomkarnBoonyaprasert replied:
flag added

boonwj commented on HoloPipelines/pipelines/components/compSeparateNumpy.py:
```diff
@@ -0,0 +1,54 @@
+            makeGlb.main(
+                pathlib.Path.cwd().joinpath("temp" + str(integer) + ".obj"),
+                outputPath.joinpath("organNo" + str(integer) + ".glb"),
+                True,
```
same here use flagName=True

UdomkarnBoonyaprasert replied:
flag added

boonwj commented on HoloPipelines/pipelines/pipelineAbdomenDicom2glb.py:
```diff
@@ -0,0 +1,26 @@
+    segmentedNiftiPath = compHttpRequest.sendFilePostRequest(
+        "http://localhost:5000/model",
+        generatedNiftiPath,
+        "_tempAbdomenSegmented.nii.gz",
```
Think about how to handle all the residue files leftover from the processes

UdomkarnBoonyaprasert replied:
nii file now gets deleted in compNifti2numpy if stated (`deleteNiftiWhenDone` argument in compNifti2numpy. The argument is set to False by default)

boonwj commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -1,5 +1,5 @@
 from components import compCommonPath
-from components import compDcm2nifti
+from components import compDicom2nifti
 from components.lungSegment.main import main as lungSegment
```
this is weird

UdomkarnBoonyaprasert replied:
import structure updated. tyty

UdomkarnBoonyaprasert committed:
PIPELINE: address PR comments
- fix comment typo in compDicom2numpy
- use `with` to open files in compHttpRequest
- add error handling in compHttpRequest
- piplineAbdomenDicom2glb now takes model URL instead of hard code
- add return code to error msg in compHttpRequest
- fix variable name typo in compNumpyTransformation
- rename crop() to res() in compNumpyTransformation
- remove resize() in compNumpyTransformation
- update the way 5D array is handled in compNumpyTransformation
- compSeparateNumpy now only takes numpy array instead of file path to npy
- add flag for arguments in compSeparateNumpy
- add function in compNifti2numpy to delete nii file if stated
- change import structure in pipelineLungDicom2glb

UdomkarnBoonyaprasert committed:
PIPELINE: fix error handling in post request
fix error handling in compHttpRequest

UdomkarnBoonyaprasert committed:
PIPELINE: minor edit for error msg

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,23 @@
+    with open(inputFile, "rb") as inputFileData:
+        file = {"file": inputFileData}
+        try:
+            response = requests.post(url, files=file, timeout=10)
```
Add validation for the response status code

UdomkarnBoonyaprasert replied:
add check for code 200

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -79,10 +79,19 @@ def resample(dataPath, new_spacing=[1, 1, 1]):
+
+
 def main(dicomPath):
     print("dicom2numpy: resampling dicom...")
```
change all print to logging.info or something.

UdomkarnBoonyaprasert replied:
update print too logging.info and logging.warning

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,49 @@
+import sys
+
+
+def res(img, newX, newY, newZ):
```
center_crop please

UdomkarnBoonyaprasert replied:
renamed

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -0,0 +1,49 @@
+        if highest > limit:
+            img = res(img, limit, limit, limit)
+
+        print("compNumpyTransformation: array resize done")
```
use logging.info

UdomkarnBoonyaprasert replied:
updated to logging

boonwj commented on HoloPipelines/pipelines/pipelineAbdomenDicom2glb.py:
```diff
@@ -0,0 +1,24 @@
+        segmentationModelUrl, generatedNiftiPath, "_tempAbdomenSegmented.nii.gz"
+    )
+    generatedNumpyList = compNifti2numpy.main(segmentedNiftiPath)
+    resizedNumpyList = compNumpyTransformation.sizeLimit(generatedNumpyList, 300)
```
Set 300 as the default value.  But leave it configurable as an argument in main()

UdomkarnBoonyaprasert replied:
add 300 as default

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,23 @@
+    inputFile = str(pathlib.Path(inputFile))
+    outputFile = str(pathlib.Path(outputFile))
+    with open(inputFile, "rb") as inputFileData:
+        file = {"file": inputFileData}
```
change variable name "file" it's somewhat of a restricted keyword

UdomkarnBoonyaprasert replied:
renamed to `fileToSend`

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,23 @@
+            response = requests.post(url, files=file, timeout=10)
+        except Exception:
+            sys.exit(
+                "compHttpReuest: an error happened in a POST reuqest, this might be due to timeout or bad request. Or if this is a request to one of the segmentation model, then please make sure the container with the model is running."
```
compHttpReuest

boonwj replied:
reuqest

UdomkarnBoonyaprasert replied:
typos fixed

UdomkarnBoonyaprasert committed:
PIPELINE: address comments from PR
- replace print() with logging
- add error handling in compHttpRequest (only accept status code 200)
- fix typos
- renamed res() to centerCrop() in compNumpytransformation

UdomkarnBoonyaprasert committed:
PIPELINE: add testing for abdomen pipeline
- add testing for abdomen pipeline
 - using mock http server within testing instead of the NN moidel container

UdomkarnBoonyaprasert committed:
PIPELINE: fix issue from codeFactor

UdomkarnBoonyaprasert committed:
PIPELINE: fix issues from CodeFactor

UdomkarnBoonyaprasert committed:
PIPELINE: update test resources url

UdomkarnBoonyaprasert committed:
PIPELINE: fix bug in pipelines test
- url for nii wasn't defined

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -59,8 +59,8 @@ def resample(dataPath, new_spacing=[1, 1, 1]):
     except Exception as e:
-        print(len(scan[0].PixelSpacing))
-        print(
+        logging.info(len(scan[0].PixelSpacing))
```
This error message can be more meaningful, so that the value makes sense to the user when they are reading the print outs on the screen.

UdomkarnBoonyaprasert replied:
error msg updated

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -1,23 +1,30 @@
         except Exception:
             sys.exit(
-                "compHttpReuest: an error happened in a POST reuqest, this might be due to timeout or bad request. Or if this is a request to one of the segmentation model, then please make sure the container with the model is running."
+                "compHttpRuest: an error happened in a POST request, this might be due to timeout or bad request. Or if this is a request to one of the segmentation model, then please make sure the container with the model is running."
```
compHttpRuest

UdomkarnBoonyaprasert replied:
typo fixed (last time, hopefully)

boonwj commented on HoloPipelines/pipelines/components/compNifti2numpy.py:
```diff
@@ -44,4 +45,4 @@ def main(inputNiftiPath, deleteNiftiWhenDone=False):
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.warning("component can't run on its own")
```
This is an error instead of a warning. Warnings indicate that the program can still be executed, which is different for this case.

UdomkarnBoonyaprasert replied:
warning changed to error for all components

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -88,12 +88,12 @@ def reorientateNumpyList(numpyList):
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.warning("component can't run on its own")
```
This is an error instead of a warning. Warnings indicate that the program can still be executed, which is different for this case.

UdomkarnBoonyaprasert replied:
warning changed to error for all components

boonwj commented on HoloPipelines/pipelines/components/compDicom2nifti.py:
```diff
@@ -15,4 +16,4 @@ def main(dicomInputPath, niftiOutputPath):
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.warning("component can't run on its own")
```
This is an error instead of a warning. Warnings indicate that the program can still be executed, which is different for this case.

UdomkarnBoonyaprasert replied:
warning changed to error for all components

boonwj commented on HoloPipelines/pipelines/components/compNumpy2obj.py:
```diff
@@ -54,9 +57,9 @@ def generateObj(inputNumpy, thisThreshold, outputObjPath):
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.warning("component can't run on its own")
```
This is an error instead of a warning. Warnings indicate that the program can still be executed, which is different for this case.

UdomkarnBoonyaprasert replied:
warning changed to error for all components

boonwj commented on HoloPipelines/pipelines/components/compNumpyTransformation.py:
```diff
@@ -35,15 +36,15 @@ def sizeLimit(img, limit):
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.warning("component can't run on its own")
```
This is an error instead of a warning. Warnings indicate that the program can still be executed, which is different for this case.

UdomkarnBoonyaprasert replied:
warning changed to error for all components

boonwj commented on HoloPipelines/pipelines/components/compObj2glbWrapper.py:
```diff
@@ -30,14 +31,14 @@ def main(inputObjPath, outputGlbPath, deleteOriginalObj=True, compressGlb=False)
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.info("component can't run on its own")
```
This is an error instead of an info. Info indicates that there is no problem

UdomkarnBoonyaprasert replied:
info changed to error

boonwj commented on HoloPipelines/pipelines/components/compSeparateNumpy.py:
```diff
@@ -39,4 +39,4 @@ def main(originalNumpy, outputPath):
 
 if __name__ == "__main__":
-    print("component can't run on its own")
+    logging.warning("component can't run on its own")
```
This is an error instead of a warning. Warnings indicate that the program can still be executed, which is different for this case.

UdomkarnBoonyaprasert replied:
warning changed to error for all components

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -18,35 +26,64 @@
 
-        print("Beginning nifti unzip...")
+        response = requests.get(url)
+        open(str(thisCwd.joinpath(zipFileName)), "wb+").write(response.content)
```
What is the difference between this and the other ways used above to download the file?

UdomkarnBoonyaprasert replied:
I previously changed to this as CodeFactor flagged it as insecure. The other two now updated to this way of downloading.

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -18,35 +26,64 @@
-    print("Beginning dicom sample download...")
+    logging.info("Checking for sample files...")
 
     if not os.path.exists(str(dicomPath.joinpath("3_Axial_CE"))):
```
Steps here are generally repeated just with different file names and paths. This step can be simplified with a loop through a dictionary of urls and paths.

UdomkarnBoonyaprasert replied:
Updated to loops

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -55,6 +92,35 @@ def testSetup():
+
+
 def remove3Dmodels():
     if os.path.exists(str(glbPath.joinpath("testResult0.glb"))):
```
This is too repetitive, you can use a loop through a list for this.

UdomkarnBoonyaprasert replied:
Updated to loops

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -55,6 +92,35 @@ def testSetup():
+    yield
+
+    # remove nii file
+    if os.path.exists(str(niftiPath.joinpath(segmentedAbdomenFileName))):
```
Same here, do not need to repeat yourself. Loops are your friend

UdomkarnBoonyaprasert replied:
updated to loops

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -5,6 +5,7 @@
 import subprocess
 import pathlib
 import pipelines.components.compCommonPath as plCommonPath
+import logging
```
There should be a way to set the logging level. The default level is `warning` and anything below it won't be shown. 

I would recommend just adding `logging.basicConfig(level=logging.INFO)` for now. Then a subsequent PR to allow log levels to be changed at runtime.

ref: https://realpython.com/python-logging/

UdomkarnBoonyaprasert replied:
added `logging.basicConfig(level=logging.INFO)` for all files with logging

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -55,6 +92,35 @@ def testSetup():
+        "https://holoblob.blob.core.windows.net/mock-pacs/Owenpap___niftynet_out.nii.gz"
+    )
+    response = requests.get(urlNiftyOut)
+    open(str(niftiPath.joinpath(segmentedAbdomenFileName)), "wb+").write(
```
Is there a need to keep downloading this file on every test run? Could it be done at the same stage with all the file downloads step.

UdomkarnBoonyaprasert replied:
Moved to `testSetup()` (this is not in the dict for sample files as it's the only one not zipped)

boonwj reviewed (changes requested):
Good job with using the test server to mock the responses. Some other changes before we can merge this.

- The test setup can be simplified with the repetition
- Logging inconsistencies
- More meaningful logging messages
- Setting initial log level so that INFO messages will be printed

boonwj commented on HoloPipelines/pipelines/pipelineAbdomenDicom2glb.py:
```diff
@@ -0,0 +1,31 @@
+
+
+if __name__ == "__main__":
+    main(sys.argv[1], sys.argv[2], sys.argv[3])
```
There's no way to change the `resolutionLimit` with this method call.

UdomkarnBoonyaprasert replied:
added extra `sys.argv` and updated pipelineList to accept 4 args

boonwj commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,30 @@
+            response = requests.post(url, files=fileToSend, timeout=10)
+            if response.status_code != 200:
+                sys.exit(
+                    "compHttpRequest: error, bad status code (got {})".format(
```
Is there any other information that will be useful when such a situation occurs? Does the server respond with anything useful in the response body that could be printed as well?

UdomkarnBoonyaprasert replied:
Updated error msg with content of the response from the server (the other status code aside from 200 is 400 with "No file in the request" or "No selected file" as the content)

boonwj commented on HoloPipelines/pipelines/pipelineAbdomenDicom2glb.py:
```diff
@@ -0,0 +1,31 @@
+def main(
+    inputDicomPath, outputGlbFolderPath, segmentationModelUrl, resolutionLimit=300
+):
+    generatedNiftiPath = compDicom2nifti.main(inputDicomPath, "_temp.nii")
```
This file `_temp.nii` is left in the base directory of the project. Do keep all temp files out of the base directory of the project. Also it's best they are cleaned up when unneeded.

UdomkarnBoonyaprasert replied:
Changed from base dir to `medicalScans/nifti`. We plan to do clean up in the PIPELINE/API branch as Niel has mentioned in one of the PR comment for API, where we will have all the downloaded and temp files in a folder the job is allocated to i.e. `workSpace/<job id>/nifti/_temp.nii` (name not final). With this we can just delete the whole `<job id>` dir once job is finished (or failed).

UdomkarnBoonyaprasert committed:
PIPELINE: address issues from PR comments
- update dupilcated codes with loops (inside pipelines test)
- fix typos
- update path in abdomen pipeline

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -47,7 +47,7 @@ def loadPixelArray(inputPath):
     scan = loadScan(dataPath)
     image = loadPixelArray(dataPath)
-    print("Shape before resampling\t", image.shape)
+    logging.info("Shape before resampling\t", image.shape)
```
Just found out that this generates the following error when used. 

```
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib64/python3.7/logging/__init__.py", line 1034, in emit
    msg = self.format(record)
  File "/usr/lib64/python3.7/logging/__init__.py", line 880, in format
    return fmt.format(record)
  File "/usr/lib64/python3.7/logging/__init__.py", line 619, in format
    record.message = record.getMessage()
  File "/usr/lib64/python3.7/logging/__init__.py", line 380, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "pipelines/pipelineDicom2glb.py", line 31, in <module>
    main(sys.argv[1], sys.argv[2], sys.argv[3])
  File "pipelines/pipelineDicom2glb.py", line 13, in main
    generatedNumpyList = compDicom2numpy.main(str(pathlib.Path(dicomFolderPath)))
  File "/home/boonwenjie/workspace/COMP0111_SUMMER_PROJ/holo-repo-clean/HoloPipelines/pipelines/components/compDicom2numpy.py", line 97, in main
    imgs_after_resamp, spacing = resample(dicomPath)
  File "/home/boonwenjie/workspace/COMP0111_SUMMER_PROJ/holo-repo-clean/HoloPipelines/pipelines/components/compDicom2numpy.py", line 52, in resample
    logging.info("Shape before resampling\t", image.shape)
Message: 'Shape before resampling\t'
Arguments: ((160, 512, 512),)
```

boonwj replied:
A fix will be to use the repr version of `image.shape` and form the log message. 
```suggestion
    logging.info("Shape before resampling\t" +  repr(image.shape))
```

boonwj replied:
This logging style is found 4 times in the project. Please change them all. 
```
./pipelines/components/compNumpy2obj.py:    logging.info(image.shape)
./pipelines/components/compDicom2numpy.py:    logging.info("Shape before resampling\t" + repr(image.shape))
./pipelines/components/compDicom2numpy.py:    logging.info("Shape after resampling\t" + repr(image.shape))
./pipelines/components/compNifti2numpy.py:    logging.info("Shape after resampling\t", image.shape[:3])
```

UdomkarnBoonyaprasert replied:
Update all `image.shape` with repr version. (removed `logging.info(image.shape)` from compNumpy2obj)

UdomkarnBoonyaprasert committed:
PIPELINE: fix issue with logging image.shape

UdomkarnBoonyaprasert committed:
PIPELINE: minor fix in logging

UdomkarnBoonyaprasert committed:
PIPELINE: fix minor issue with logging

nielsboecker reviewed (approved):
Thanks for your work on this, guys.

One thing to keep in mind is that we still don't have one output that can be accepted by the HoloStorage for this pipeline, should be cleared in a new PR. See #48
